cmd: use a preinit_array function rather than parsing /proc/self/cmdline

jhenstridge opened the pull request:
At present, the C bootstrap code in snap-update-ns tries to parse `/proc/self/cmdline` because constructor functions are not given access to the argv vector passed to `main`.

If we instead add a function pointer to the `.preinit_array` ELF section, glibc will call that function with a standard argv vector:

https://sourceware.org/git/?p=glibc.git;a=blob;f=csu/elf-init.c;h=34ee89b843e966723a645c7142b7983682c33919;hb=HEAD#l66

We can then remove the code that tries to decode the information from `/proc`.

The ELF standard also defines a `.init_array` section, which I chose not to use: it seems that something else (possibly Go itself?) was adding an entry to `.init_array`, so this ensures our bootstrap code runs first.

mvo5 commented:
Woah, I really like this PR!

zyga commented:
Really really nice! I when I was doing this initially I was hoping that the argv array is (perhaps) somewhere in the initialization vector, available by some funky sycall or other mechanism but I could not find one.

As a silly remark, `init`, `preinit` is there also `prepreinit`? :-)

jhenstridge commented:
The main reason I refactored that code into a `has_option` function was as a starting point for integrating the user mount code into snap-update-ns.

As for handling unknown options, it looks like the code currently in master only causes an error if the _first_ option is unknown.  If I pass a second unknown option, the bootstrap code will execute just fine and only stop when Go attempts to parse the command line arguments.

Do we have a good idea of what we're trying to protect against here?

zyga commented:
+1 on the rationale @jhenstridge

jdstrand commented:
snap-update-ns is now called by the setuid snap-confine which can be invoked by the user. The rationale for the bootstrap code is that it is supposed to perform all input validation and provide a safe environment (and fail early) such that the go code can just run with it such that the go code can leverage whatever imports it wants without worrying about how the inputs and environment might affect it. If there are bugs in this part of the code, we should definitely fix them.

jhenstridge commented:
I've updated `process_arguments` to handle the missing checks I mentioned earlier (multiple positional arguments, invalid option after the first), and reinstated the relevant checks.

It was easier to refactor the code to scan the argument list once, so I got rid of the helpers and their associated tests.  I added extra tests for `process_arguments` to compensate.

mvo5 commented on cmd/snap-update-ns/bootstrap.go:
```diff
@@ -69,47 +67,38 @@ func BootstrapError() error {
 // This also sets bootstrap_msg on failure.
 func validateSnapName(snapName string) int {
 	cStr := C.CString(snapName)
+	defer C.free(unsafe.Pointer(cStr))
```
Nice catch

bboozzoo commented:
Probably not a huge problem, but this will not work under musl. So if anyone would want to build a minimal, boot-to-snapd system, they'll have to fix this themselves.

jhenstridge commented:
We already have a minimal "boot to snapd" system in the form of Ubuntu Core, where snap-update-ns would be run from within the core snap.  And while there are plans to shrink the core snap down via the introduction of base snaps, I don't think there is a plan to get rid of glibc.

zyga commented on cmd/snap-update-ns/bootstrap.go:
```diff
@@ -28,9 +28,7 @@ package main
-__attribute__((constructor)) static void init(void) {
-	bootstrap();
-}
+__attribute__((section(".preinit_array"), used)) static typeof(&bootstrap) init = &bootstrap;
```
Could you please add a comment explaining how this works:
```
The bootstrap function is called by the runtime before passing control to main.
We are using the `preinit_array` instead of `init_array` in case go later on decides
to use the latter for any initialization which would defeat the purpose of the C
bootstrap code.
```

jdstrand commented on cmd/snap-update-ns/bootstrap.go:
```diff
@@ -69,47 +67,38 @@ func BootstrapError() error {
-		str := C.GoString(ptr)
-		return &str
+func makeArgv(args []string) []*C.char {
+	argv := make([]*C.char, len(args)+1)
```
For future clarify, can you add a comment:
```
// Create argv array with a terminating NULL element
```

jdstrand commented on cmd/snap-update-ns/bootstrap.go:
```diff
@@ -69,47 +67,38 @@ func BootstrapError() error {
-		return &str
+func freeArgv(argv []*C.char) {
+	for _, arg := range argv {
+		C.free(unsafe.Pointer(arg))
```
Technically we don't have to free the last element, but free(NULL) is a no-op so ok.

jdstrand commented on cmd/snap-update-ns/bootstrap_test.go:
```diff
@@ -31,134 +29,50 @@ type bootstrapSuite struct{}
 		// The option --from-snap-confine disables setns.
-		{"argv0\x00--from-snap-confine\x00snapname\x00", "snapname", false, ""},
+		{[]string{"argv0", "--from-snap-confine", "snapname"}, "snapname", false, ""},
+		{[]string{"argv0", "snapname", "--from-snap-confine"}, "snapname", false, ""},
```
This is different from the original design (<name> followed by <option>), but I think that is ok.

jhenstridge replied:
The existing code seems to allow either order (as does the Go argument parsing code), so I went ahead and added the extra test.  We could turn it into an error, but that'd likely make the code a bit more complicated.

jdstrand replied:
Right, that's precisely why I don't think it is important (adding code complexity). IMO, don't worry about it.
